docs/src/test/java/docs/user/effects/ShoppingCartEntity.java

```diff
@@ -0,0 +1,58 @@
+package docs.user.effects;
+
+import com.example.Hotitems;
+import com.example.Shoppingcart;
+import com.google.protobuf.Empty;
+import io.cloudstate.javasupport.Context;
+import io.cloudstate.javasupport.ServiceCallRef;
+import io.cloudstate.javasupport.eventsourced.CommandContext;
+import io.cloudstate.javasupport.eventsourced.CommandHandler;
+
+public class ShoppingCartEntity {
+
+  // #lookup
+  private final ServiceCallRef<Hotitems.Item> itemAddedToCartRef;
+
+  public ShoppingCartEntity(Context ctx) {
+    itemAddedToCartRef =
+        ctx.serviceCallFactory()
+            .lookup("example.shoppingcart", "ItemAddedToCart", Hotitems.Item.class);
```

Comment:
If this method requires having the proto classes for the endpoint-to-be-called, is there any other way of referring to an endpoint rather than look up by Strings?

Reply:
No. Not unless you've compiled the grpc descriptor to something. We don't require the gRPC descriptor to be compiled to anything in Java, since we dynamically serve it, we just read the descriptor directly, and then users implement it using annotations which are reflected over. By default, nothing will be generated for the gRPC services, other than the descriptor, and so they can only be looked up by string. If we added support for generating code, then we could in fact remove the need for the `ServiceCallFactory` at all, and instead emitting an effect could look like this:

```
ctx.effect(HotItemsService.ref.itemAddedToCart(item));
```